use-cases/ai-agent.ipynb

```diff
@@ -0,0 +1,560 @@
+    "\n",
+    "    # Initialize a list of inputs to pass to the LLM\n",
+    "    llm_input = []\n",
+    "\n",
+    "    # Retrieve the session history for the current session and add it to the LLM input\n",
+    "    session_history = retrieve_session_history(session_id)\n",
+    "    llm_input.extend(session_history)\n",
+    "\n",
+    "    # Append the user message in the correct format\n",
+    "    user_message = {\n",
+    "        \"role\": \"user\",\n",
+    "        \"content\": user_input\n",
+    "    }\n",
+    "    llm_input.append(user_message)\n",
+    "\n",
+    "    # Call the tool_selector function to determine which tool to use\n",
+    "    tool, tool_input = tool_selector(user_input, session_history)\n",
+    "    print(\"Tool selected: \", tool)\n",
+    "    \n",
+    "    # Process based on selected tool\n",
+    "    if tool == \"vector_search_tool\":\n",
+    "        context = vector_search_tool(tool_input)\n",
+    "        # Construct the system prompt using the retrieved context and append it to the LLM input\n",
+    "        system_message_content = (\n",
+    "            f\"Answer the user's question based on the retrieved context and conversation history.\\n\"\n",
+    "            f\"1. First, understand what specific information the user is requesting\\n\" \n",
+    "            f\"2. Then, locate the most relevant details in the context provided\\n\"\n",
+    "            f\"3. Finally, provide a clear, accurate response that directly addresses the question\\n\\n\"\n",
+    "            f\"If the current question builds on previous exchanges, maintain continuity in your answer.\\n\"\n",
+    "            f\"Only state facts clearly supported by the provided context. If information is not available, say 'I DON'T KNOW'.\\n\\n\"\n",
+    "            f\"Context:\\n{context}\"\n",
+    "        )\n",
+    "        response = get_llm_response(llm_input, system_message_content)\n",
+    "    elif tool == \"calculator_tool\":\n",
+    "        # Perform the calculation using the calculator tool\n",
+    "        response = calculator_tool(tool_input)\n",
+    "    else:\n",
+    "        system_message_content = \"You are a helpful assistant. Respond to the user's prompt as best as you can based on the conversation history.\"\n",
+    "        response = get_llm_response(llm_input, system_message_content)\n",
+    "    \n",
+    "    # Store the system response in the chat history collection\n",
+    "    store_chat_message(session_id, \"system\", response)\n",
+    "    return response"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Test the agent\n",
+    "Run a few queries to test the agent. The agent determines the best tool to use based on the query and retains memory from previous interactions to inform future responses."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [
```

Comment:
Since you have consistently provided the outputs, I'm guessing it was intentional, but just flagging to follow up in case.

For this particular cell, it was obvious that it was executing because it took time to perform the workflow. I could see a partial response while the next Q was being answered.

For some of the cells above where we provided output, when I executed the cell that was supposed to _create_ the output, it wasn't obvious that it had executed correctly because it returned so quickly I didn't really see the output change/get generated. I generally appreciate that the output confirms that the cell has executed correctly, but when the output is already present, it's less obvious that the cell has executed correctly. Just my $0.02.

Reply:
I see where you're coming from. My rationale for including the output for this notebook as opposed to others is that for longer notebooks, or for tutorials that might not be easily runnable (e.g. no access to Voyage AI key), users can at least see the expected output/workflow. I'm inclined to keep them for that reason!